include/boost/beast/core/impl/saved_handler.hpp

```diff
@@ -0,0 +1,193 @@
+//
+// Copyright (c) 2016-2017 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef BOOST_BEAST_CORE_IMPL_SAVED_HANDLER_HPP
+#define BOOST_BEAST_CORE_IMPL_SAVED_HANDLER_HPP
+
+#include <boost/beast/core/detail/allocator.hpp>
+#include <boost/asio/associated_allocator.hpp>
+#include <boost/asio/associated_executor.hpp>
+#include <boost/asio/executor_work_guard.hpp>
+#include <boost/assert.hpp>
+#include <boost/core/empty_value.hpp>
+#include <boost/core/exchange.hpp>
+#include <memory>
+#include <utility>
+
+namespace boost {
+namespace beast {
+
+//------------------------------------------------------------------------------
+
+class saved_handler::base
+{
+protected:
+    ~base() = default;
+
+public:
+    base() = default;
+    virtual void destroy() = 0;
+    virtual void invoke() = 0;
+};
+
+//------------------------------------------------------------------------------
+
+template<class Handler, class Alloc>
+class saved_handler::impl final : public base
+{
+    using alloc_type = typename
+        beast::detail::allocator_traits<
+            Alloc>::template rebind_alloc<impl>;
+
+    using alloc_traits =
+        beast::detail::allocator_traits<alloc_type>;
+
+    struct ebo_pair : boost::empty_value<alloc_type>
+    {
+        Handler h;
+
+        template<class Handler_>
+        ebo_pair(
+            alloc_type const& a,
+            Handler_&& h_)
+            : boost::empty_value<alloc_type>(
+                boost::empty_init_t{}, a)
+            , h(std::forward<Handler_>(h_))
+        {
+        }
+    };
+
+    ebo_pair v_;
+    net::executor_work_guard<
+        net::associated_executor_t<Handler>> wg_;
+
+public:
+    template<class Handler_>
+    impl(alloc_type const& a, Handler_&& h)
+        : v_(a, std::forward<Handler_>(h))
+        , wg_(net::get_associated_executor(v_.h))
```

Comment:
This breaks composed operations that use the non-intrusive executor hook, because if the operation has no associated executor, then this will select `system_executor` rather than the I/O executor.

Reply:
I believe that should be okay, because the caller is already required to maintain a work guard for the operation which "owns" this saved handler?

Reply:
If a user writes an async algorithm that performs a single operation, there's no need to maintain the work guard.

Reply:
the previous version had the guard, you think it wasn't needed?

Reply:
I'm not saying the guard is not needed, just that the executor is not selected properly. [async.reqmts.async.work] indicates we need guards for `ex1` and `ex2`, but here, `ex1` is implicitly somewhere else, so we don't actually need it. `ex2` is `get_associated_executor(completion_handler,
ex1)`